src/coreComponents/functions/MultivariableTableFunctionKernels.hpp

```diff
@@ -0,0 +1,321 @@
+ */
+template< integer NUM_DIMS, integer NUM_OPS >
+class MultivariableTableFunctionStaticKernel
+{
+public:
+
+  /// Compile time value for the number of table dimensions (inputs)
+  static constexpr integer numDims = NUM_DIMS;
+
+  /// Compile time value for the number of operators (interpolated functions, outputs)
+  static constexpr integer numOps = NUM_OPS;
+
+  /**
+   * @brief Construct a new Multivariable Table Function Static Kernel object
+   *
+   * @param[in] axisMinimums  minimum coordinate for each axis
+   * @param[in] axisMaximums maximum coordinate for each axis
+   * @param[in] axisPoints number of discretization points between minimum and maximum for each axis
+   * @param[in] axisSteps axis interval lengths (axes are discretized uniformly)
+   * @param[in] axisStepInvs inversions of axis interval lengths (axes are discretized uniformly)
+   * @param[in] axisHypercubeMults  hypercube index mult factors for each axis
+   * @param[in] hypercubeData table data stored per hypercube
+   * @param[in] coordinates array of coordinates of points where interpolation is required
+   * @param[in] values array of interpolated operator values (all operators are interpolated at each point)
+   * @param[in] derivatives  array of derivatives of interpolated operators (all derivatives for all operators are computed at each point)
+   */
+  MultivariableTableFunctionStaticKernel( arrayView1d< real64 const > const & axisMinimums,
+                                          arrayView1d< real64 const > const & axisMaximums,
+                                          arrayView1d< integer const > const & axisPoints,
+                                          arrayView1d< real64 const > const & axisSteps,
+                                          arrayView1d< real64 const > const & axisStepInvs,
+                                          arrayView1d< globalIndex const > const & axisHypercubeMults,
+                                          arrayView1d< real64 const > const & hypercubeData,
+                                          arrayView1d< real64 const > const & coordinates,
+                                          arrayView1d< real64 > const & values,
+                                          arrayView1d< real64 > const & derivatives ):
+    m_axisMinimums ( axisMinimums ),
+    m_axisMaximums ( axisMaximums ),
+    m_axisPoints ( axisPoints ),
+    m_axisSteps ( axisSteps ),
+    m_axisStepInvs ( axisStepInvs ),
+    m_axisHypercubeMults ( axisHypercubeMults ),
+    m_hypercubeData ( hypercubeData ),
+    m_coordinates ( coordinates ),
+    m_values ( values ),
+    m_derivatives ( derivatives )
+  {};
+
+  /**
+   * @brief Compute interpolation element-wise using host or device execution space
+   *
+   * @param[in] ei Index of element
+   */
+  GEOSX_HOST_DEVICE
+  void
+  compute( localIndex const ei ) const
+  {
+    interpolatePoint( &m_coordinates[ei * NUM_DIMS], &m_values[ei * NUM_OPS], &m_derivatives[ei * NUM_OPS * NUM_DIMS] );
+  }
```

Comment:
I am not convinced that this is the most convenient interface for this kernel. I see two main options (which are not mutually exclusive):

1. If it is to remain a generic utility kernel interface for multivalue interpolation under `functions`, it should look more like `TableFunction::KernelWrapper::compute()`, i.e. take both its inputs and outputs (values for a single interpolation point) as arguments. This makes it more flexible and does not bind the client code to storing values in global memory or using a particular type for storage (`array1d` in this case). For example, someone could pass in pointers to local stack-based arrays if interpolation is fused with some other computation or assembly process and the interpolation results are only needed as intermediate values.
2. If we want a kernel in the style of `CompositionalMultiphaseBaseKernels`, i.e. something that is particular to a specific choice of variables and storage scheme, it should also take care of extracting its inputs/outputs from data repository and live under `physicsSolvers`, accompanying the corresponding OBL-based flow solver.

I'm fine with merging things as is and waiting until we get a usage example from the actual solver before committing to a particular choice.

Reply:
I completely agree, @klevzoff, this is not meant to be the final interface for this kernel! I would most probably go with the first option, where the in-place interpolation you described fits perfectly for the (second iteration of) very same OBL solver: it would save a lot of intermediate space (and maybe even increase performance). I am going to use such space for the first implementation to store interpolated operator values and derivatives - it is the same (and only) way we have it in DARTS. Because GEOSX kernel design is much more generic and flexible, it`ll be very straightforward to test the in-place interpolation here.